### PR TITLE
Changed to Typescript. Added SparseTypedFastBitSet

### PR DESCRIPTION
**Pull request by ashley-taylor**
Apologies for this very large PR.

We make heavy use of this library.
Our use-case is similar to maintaining relationships between nodes in a graph database

Some nodes have a huge number of relationships and others very few. Need often perform union/intersections etc between these relationships.

We noticed that memory and iteration speed over these very sparse relationships was very inefficient.

Adding this Sparse implementation has taken our memory usage from 600MB of bitset information to about 20MB
And iteration performance is hugely improved. Went from the bulk of the time reporting in the bitset code to hard to identify impact.

Changes.
Changed the TypedFastBitSet to a Typescript class. 
This is the identical code to the javascript one.
I ran the benchmark from the other Fastbitset project. And they seem to be mostly comparable in performance.
Typescript version seems slightly quicker think due to not double allocation the `Uint32Array` in a few places and storing things like the array size as a local variable might improve the JIT in a few cases.

Unfortunately how it was exported does mean that there is a change when upgrading to this version.
Changed from
```
const TypedFastBitSet = require("typedfastbitset");
```
To
```
const { TypedFastBitSet } = require("typedfastbitset");
```

Added an interface that both classes implement and can be used together interchangeably 

New `SparseTypedFastBitSet`
Tried to make memory usage as small as possible. Only added a single extra number to the class `arraySize`. 
if `arraySize` is -1 then is a bitset
Then for the first 256 entries, they are added to an array.

The first entry is always the largest value. Used so when switching to a bitset know the correct size.

Added tests over the interface that test both implementations.

To release assuming npm keys etc are set up on your machine should be able to run

`npm run release`

Happy to make any changes you see fit

### CODE REVIEW

File: src/SparseTypedFastBitSet.ts [hunk 1]
```diff
@@ -0,0 +1,1380 @@
...
+      if (words.length << 5 > index) return Type.BITSET;
+      const count = (index + 32) >>> 5; // just what is needed
+      const newwords = new Uint32Array(count << 1);
+      newwords.set(words); // hopefully, this copy is fast
+      this.data = newwords;
+
+      return Type.BITSET;
+    } else {
+      const array = this.data;
+      if (this.arraySize + capacity < array.length) {
```
Comment (lemire) [inline src/SparseTypedFastBitSet.ts:432]: So irrespective of 'number', if I call resize(number) and arraySize is smaller than array.length, then this function returns Type.ARRAY and does nothing else. Is that correct?

**Comment by lemire** [comment]
@ashley-taylor

Because you have rewritten everything, we need to assess performance.

Please sync with the main branch, as I have added back the benchmarking script.

On your favorite system, please run the following from our main branch:

```
cd benchmark
npm install --no-save benchmark microtime bitset.js fastbitset fast-bitset roaring
nodejs test.js 
```

Please update to your pull request and run the same command. Evidently, you may need to make the change you suggest (`const { TypedFastBitSet } = require("typedfastbitset");`), other than that **please do not change the benchmark**, we want to run the same benchmarking code before and after your PR.

File: src/SparseTypedFastBitSet.ts (hunk 1 above)
Comment (ashley-taylor) [reply src/SparseTypedFastBitSet.ts:432]: Yip method returns the current datastructure used.

Save the calling method from doing the same check itself.

So if is basic array and capacity will return ARRAY if is a bitset and doesn't need to resize will return BITSET

**Comment by lemire** [comment]
@ashley-taylor If you could publish your benchmark results, that would be great.

**Comment by ashley-taylor** [comment]
Have run the benchmark.
Have to use quite an old version of node because of roaring being compiled against a version.

[baseline.txt](https://github.com/lemire/TypedFastBitSet.js/files/8578258/baseline.txt) running the test suite on the existing master
[rewrite.txt](https://github.com/lemire/TypedFastBitSet.js/files/8578262/rewrite.txt)
Just changing the import on the test suite to point to new code

[sparse.txt](https://github.com/lemire/TypedFastBitSet.js/files/8578266/sparse.txt) added sparse implementation to benchmark


[merged.txt](https://github.com/lemire/TypedFastBitSet.js/files/8578268/merged.txt) merged sparse in baseline so can be read side by side

**Comment by lemire** [comment]
The benchmarks appear not to reveal any problem. That's good. Thanks for running that.

**Comment by lemire** [comment]
We are going to need extra eyes on this.

**Comment by Codex-** [comment]
I'm happy to have a look, I'll get on to it in the next day or so :)

File: src/TypedFastBitSet.ts [hunk 2]
```diff
@@ -0,0 +1,593 @@
...
+ *  c.difference(b); // from c, remove elements that are in b (modifies c)
+ *  c.difference2(b); // from c, remove elements that are in b (modifies b)
+ *  c.change(b); // c will contain elements that are in b or in c, but not both
+ *  const su = c.union_size(b);// compute the size of the union (bitsets are unchanged)
+ *  c.union(b); // c will contain all elements that are in c and b
+ *  const s1 = c.intersection_size(b);// compute the size of the intersection (bitsets are unchanged)
+ *  c.intersection(b); // c will only contain elements that are in both c and b
+ *  c = b.clone(); // create a (deep) copy of b and assign it to c.
+ *  c.equals(b); // check whether c and b are equal
+
```
Comment (Codex-) [inline src/TypedFastBitSet.ts]: nit: missing `*`

File: src/TypedFastBitSet.ts [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+function isIterable(
+  obj: Iterable<number> | null | undefined
+): obj is Iterable<number> {
+  if (obj) {
+    return obj[Symbol.iterator] !== undefined;
+  }
+  return false;
+}
+// you can provide an iterable
+// an exception is thrown if typed arrays are not supported
```
Comment (Codex-) [inline src/TypedFastBitSet.ts]: Should this be a class docstring?
```ts
/**
 * You can provide an iterable an exception is thrown if
 * typed arrays are not supported.
 */
export class TypedFastBitSet implements BitSet {
```

File: src/TypedFastBitSet.ts [hunk 4]
```diff
@@ -0,0 +1,593 @@
...
+
+export class TypedFastBitSet implements BitSet {
+  //constructor
```
Comment (Codex-) [inline src/TypedFastBitSet.ts]: Probably don't need this now

File: src/TypedFastBitSet.ts [hunk 5]
```diff
@@ -0,0 +1,593 @@
...
+  ) {
+    if (isIterable(iterable)) {
+      for (const key of iterable) {
+        this.add(key);
+      }
+    }
+  }
+
+  // Returns a new TypedFastBitset given a Uint32Array
+  // of words
```
Comment (Codex-) [inline src/TypedFastBitSet.ts]: nit: would be nice to see function comments converted to jsdoc:
```ts
  /**
   * @returns a new TypedFastBitset given a Uint32Array of words
   */
```
image

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
+name: Test
```
Comment (Codex-) [inline .github/workflows/test.yml:1]: With adding this, is `.travis.yml` still needed or used? @lemire would know 🤔

File: src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+   * Returns a new TypedFastBitset given a Uint32Array of words
+   */
+
+  /** Add the value (Set the bit at `index` to `true`) */
+  add(index: number): void;
+
+  /** Set bits from start (inclusive) to end (exclusive) */
+  addRange(start: number, end: number): void;
+
+  /** Return an array with the set bit locations (values) */
```
Comment (Codex-) [inline src/utils.ts:40]: nit: ideally use jsdoc tags to indicate `@returns` for docstrings to describe the returned value.

File: src/utils.ts [hunk 8]
```diff
@@ -0,0 +1,155 @@
+// fast function to compute the Hamming weight of a 32-bit unsigned integer
```
Comment (Codex-) [inline src/utils.ts:1]: nit: mixture of jsdoc and function comments

File: src/TypedFastBitSet.ts [hunk 9]
```diff
@@ -0,0 +1,593 @@
...
+    this.words = newwords;
+  }
+
+  // How many values stored in the set? How many set bits?
+  size() {
+    const words = this.words;
+    let answer = 0;
+    const c = words.length;
+    let k = 0 | 0;
+    for (; k + 4 < c; k += 4) {
```
Comment (Codex-) [inline src/TypedFastBitSet.ts:218]: nit: might be nicer to read if these `for (; x)` loops were `while (x)`

File: src/SparseTypedFastBitSet.ts [hunk 10]
```diff
@@ -0,0 +1,1380 @@
...
+ * Alternative version to TypedFastBitSet
+ * This starts of as storing entries as an array in a Uint32Array Typed Array
+ * The largest set value is first in the array. Everything else has no defined order
+ * The largest set value is used when converting into a bitset to correctly size new Array
+ * Once more than 256 entries have been added structure will convert to a bitset if it will create a smaller array
+ * Very sparse data will remain an array until 1024 entries where it will convert to a bitset
+ *
+ *
+ *
+ *
```
Comment (Codex-) [inline src/SparseTypedFastBitSet.ts]: nit

File: src/SparseTypedFastBitSet.ts [hunk 11]
```diff
@@ -0,0 +1,1380 @@
...
+ */
+
+import { BitSet, hammingWeight, hammingWeight4 } from "./utils";
+
+enum Type {
+  ARRAY,
+  BITSET,
+  MIXED,
+}
```
Comment (Codex-) [inline src/SparseTypedFastBitSet.ts:22]: nit: missing nl

File: tsconfig.json [hunk 12]
```diff
@@ -1,14 +1,21 @@
...
+    "downlevelIteration": true,
+    "esModuleInterop": true,
+    "moduleResolution": "node",
+    "removeComments": false,
+    "sourceMap": true,
+    "stripInternal": true
   },
-  "include": ["unit"]
+  "include": ["src/**/*"],
+  "exclude": ["src/**/*.spec.*"]
```
Comment (Codex-) [inline tsconfig.json]: There are some consequences to this, mostly to the develop experience where the settings here not are reflected in your tests and your tests in the IDE may not correctly work with the TS server.

My recommendation around this to facilitate a better dev experience is to make these changes:

`tsconfig.json`

Updated to include tests, but not to emit any code.

```json
{
  "compilerOptions": {
    "target": "es5",
    "module": "commonjs",
    "declaration": true,
    "noEmit": true,
    "pretty": true,
    "sourceMap": true,

    "outDir": "lib",
    "rootDir": "src",

    "lib": ["esnext"],

    "strict": true,
    "downlevelIteration": true,
    "esModuleInterop": true,
    "moduleResolution": "node",
    "removeComments": false,
    "stripInternal": true
  },
  "include": ["src/**/*"]
}
```

Add a new file, `tsconfig.build.json` which excludes tests from the bundle, and disables `noEmit`:
```json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "noEmit": false
  },
  "exclude": ["src/**/*.spec.*"]
}
```

Update `package.json` to reflect this with `package` being updated to `tsc -p tsconfig.build.json`

File: package.json [hunk 13]
```diff
@@ -6,27 +6,34 @@
...
   },
-  "main": "TypedFastBitSet.js",
-  "types": "TypedFastBitSet.d.ts",
+  "main": "lib/index.js",
+  "types": "lib/index.d.ts",
   "scripts": {
-    "test": "tsc && mocha -r ts-node/register unit/*"
+    "build": "tsc",
+    "test": "node --expose-gc ./node_modules/jest/bin/jest",
+    "package": "tsc -p tsconfig.json",
```
Comment (Codex-) [inline package.json]: with tsconfig changes: `tsc -p tsconfig.build.json`, i'd consider making this `build:package`.

File: package.json [hunk 14]
```diff
@@ -6,27 +6,34 @@
...
+    "release": "tsc -p tsconfig.json && node --expose-gc ./node_modules/jest/bin/jest && release-it"
```
Comment (Codex-) [inline package.json]: can be `npm run test && npm run package && release-it`, although should `release-it`

File: package.json [hunk 15]
```diff
@@ -6,27 +6,34 @@
...
-    "benchmark": "~1.0.0",
-    "mocha": "^8.2.1",
-    "ts-node": "^9.1.1",
-    "typescript": "^4.2.4"
+    "@types/jest": "^27.4.1",
+    "jest": "^27.5.1",
+    "ts-jest": "^27.1.4",
+    "ts-node": "^10.7.0",
+    "typescript": "^4.6.3",
+    "release-it": "^14.14.2"
```
Comment (Codex-) [inline package.json]: Should `release-it` be added? Can be used to release and publish to npm automatically (can provide examples of this if the author wants) but I think adding releasing tooling should probably be up to @lemire ?

File: benchmark/test.js [hunk 16]
```diff
@@ -4,7 +4,7 @@
 "use strict";
```
Comment (Codex-) [inline benchmark/test.js:4]: could add a command for this into the `package` scripts? `npm run package && node ./benchmark/test.js`

**Review by Codex- — COMMENTED** [review]
Looks pretty good overall, only a few nits really and some suggestions to improve how tsconfig is used. 

Great test coverage and ideas for testing against both bitset types, coverage looks good:
image

**Comment by Codex-** [comment]
@lemire we've done some tidying up, would be great if you could take another look :)

**Comment by lemire** [comment]
@Codex- This is high quality work and I think we could move forward, but I would like you to document the state in the sparse bitset (MIXED, ARRAY, BITSET). I think it should be explicit what they represent, especially MIXED. We can guess from your description, but we would like to have specific statements... otherwise maintaining this code might prove difficult.

**Comment by lemire** [comment]
I am a bit uneasy, but ok, let us merge and see what happens.

**Pull request merged**